* [ ] check that primers cover as many of the 545 genes suggested for cgMLST by [Guglielmini et al. 2019](https://doi.org/10.1371/journal.pntd.0007374) as possible
* [ ] and the 7 genes used for traditional MLST (see [Caimi et al. 2017](https://www.sciencedirect.com/science/article/pii/S1567134817302708?via%3Dihub))
* [ ] compare sum of amplicon size from SWGA to sum of amplicon size from traditional (~7 gene) MLST schemes
* [ ] could calculate number of unique (non-overlapping) bases contributed assuming 5kb extension from each predicted touchdown location as in Coxiella paper

Questions

* Why would [Guglielmini et al. 2019](https://doi.org/10.1371/journal.pntd.0007374) remove loci from previously defined Leptospira MLST schemes? Seems like these would still be informative.

Points to consider

* Order primers with extra strong bonds on the 3' end to prohibit primer degredation leading to too many matches in background genome(s)
* May want/need to break up or use as exclusion seqs eukaryotic mitochondrial DNA to avoid rolling-circle amplification (?) due to primer match - coxiella paper did this
* What kind of samples are good candidates for rapid genomic diagnostics? Should inform target, background genome choices
* Sequence pre- and post-swga samples in order to assess how successful swga was

Supplies needed

* Primers with 3' end thiophosphate modified to withstand phi29 polymerase
* phi29 polymerase
* restriction endonuclease for mitochondrial DNA?
* A PCR purification kit that can handle really long (ideally >10000pb) amplicons
