---
title: "Assess SWGA candidate primer set for genome coverage"
output: html_notebook
---

```{r, setup}
library(dplyr)
library(ggplot2)
library(tidyr)
knitr::opts_knit$set(root.dir = '/Users/nadeau/Documents/CDC_ORISE/Projects/swga-lepto')
```

Load bedfile for each primer
```{r}
set_idx <- 6
bed_dir <- paste(
    "results",
    "lepto_human_no_downsample_fg_bind_rate0.00005_max_sets_search1000",
    "swga",
    "set_bedfiles",
    "target.fasta_export",
    sep = "/"
)
outdir <- paste0(bed_dir, "/../../../")
primer_bedfiles <- list.files(paste0(bed_dir, "set_", set_idx, "/"))
primer_bedfiles <- primer_bedfiles[primer_bedfiles != "whole_set.bed"]

is_first <- T
for (bedfile in primer_bedfiles) {
    bed <- read.delim(
        file = paste0(bed_dir, "set_", set_idx, "/", bedfile),
        skip = 1,
        header = F,
        col.names = c("ref", "primer_start", "primer_end"),
        sep = ""
    )
    bed$primer_seq <- gsub(x = bedfile, pattern = ".bed", replacement = "")
    if (is_first) {
        primer_sites <- bed
        is_first <- F
    } else {
        primer_sites <- rbind(primer_sites, bed)
    }
}
```

Report total number of (non-overlapping) sites covered per subsequent primer
And get all the sites covered
```{r}
assumed_extension = 10000
covered_sites <- c()
primers <- unique(primer_sites$primer_seq)
new_coverage_by_primer <- data.frame(
  primer_consideration_order = 1:length(primers),
  primer = primers,
  n_binding_sites = NA,
  assumed_extension = assumed_extension,
  n_new_sites_covered = NA
)

sites_covered <- c()
for (primer in primers) {
  n_new_sites_covered <- 0
  filtered_primer_sites <- primer_sites %>% filter(primer_seq == primer)
  new_coverage_by_primer[[new_coverage_by_primer$primer == primer, "n_binding_sites"]] <- nrow(filtered_primer_sites)
  for (i in 1:nrow(filtered_primer_sites)) {
    start <- filtered_primer_sites[[i, "primer_start"]]
    end <- filtered_primer_sites[[i, "primer_end"]]
    sites_covered_i <- start:(end + assumed_extension)
    n_new_sites_covered <- n_new_sites_covered + length(setdiff(sites_covered_i, sites_covered))
    sites_covered <- union(sites_covered_i, sites_covered)
  }
  new_coverage_by_primer[[new_coverage_by_primer$primer == primer, "n_new_sites_covered"]] <- n_new_sites_covered
}

print(new_coverage_by_primer)
print(paste("Total sites covered assuming touchdown at all",
            sum(new_coverage_by_primer$n_binding_sites),
            "sites and extension of",
            assumed_extension,
            "bp:",
            length(sites_covered)))

write.table(
  x = new_coverage_by_primer,
  file = paste0(
    outdir, "/genome_coverage/set_", set_idx, "_new_coverage_by_primer", assumed_extension, ".txt"
  ),
  row.names = F)
```

Get MLST loci covered
```{r}
scheme_1 <- c("glmU1", "pntA1", "sucA1", "tpiA1", "pfkB1", "mreA1", "caiB1")
scheme_2 <- c("adk2", "glmU2", "icdA2", "lipL322", "lipL412", "mreA2", "pntA2")
scheme_3 <- c("adk3", "icdA3", "lipL323", "lipL413", "rrs23", "secY3")
mlst_info <- read.table(file = "target_lepto_mlst_site_info.txt", sep = "\t", header = T)

mlst_info <- mlst_info %>% mutate(
  assumed_extension = assumed_extension,
  percent_covered = NA,
  n_sites = NA,
  mlst_scheme = case_when(
    Locus %in% scheme_1 ~ "1",
    Locus %in% scheme_2 ~ "2",
    Locus %in% scheme_3 ~ "3",
    grepl("LEPT", Locus) ~ "cgMLST",
    T ~ "Other")
)

# How many MLST loci are completely covered?
for (i in 1:nrow(mlst_info)) {
  if (i %% 100 == 0) {
    print(paste("Evaluating MLST loci", i))
  }
  start <- mlst_info[[i, "Start.position"]]
  end <- mlst_info[[i, "End.position"]]
  mlst_info[[i, "n_sites"]] <- end - start + 1
  mlst_info[[i, "percent_covered"]] <- sum(start:end %in% sites_covered) / (end - start + 1)
}

print(sum(mlst_info$percent_covered == 1, na.rm = T)/nrow(mlst_info))

ggplot(
  data = mlst_info,
  aes(x = percent_covered)) +
  geom_histogram() +
  facet_wrap(. ~ mlst_scheme, scales = "free_y")

ggsave(
    filename = paste0(
        outdir, "/mlst_loci_coverage/set_", set_idx, "_assumed_extension_", assumed_extension, ".png"
    )
)
```
